File: python/_144.py
# Solution 1. recursive
# O(N) / O(H)
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        arr = []
        def preorder(node):
            if not node:
                return
            arr.append(node.val)
            preorder(node.left)
            preorder(node.right)
        preorder(root)
        return arr


# Solution 2. iterative
# O(N) / O(H)
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        ans, stack = [], []
        while root or stack:
            if root:
                stack.append(root)
                ans.append(root.val)
                root = root.left
            else:
                root = stack.pop().right
        return ans

File: cpp/_144.cpp
// Solution 1. recursive
// O(N) / O(H)
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        preorder(root, ans);
        return ans;
    }
    
    void preorder(TreeNode* root, vector<int>& ans) {
        if (!root) return;
        ans.push_back(root->val);
        preorder(root->left, ans);
        preorder(root->right, ans);
    }
};


// Solution 2. iterative
// O(N) / O(H)
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<TreeNode*> s;
        while (root || !s.empty()) {
            if (root) {
                s.push(root);
                ans.push_back(root->val);
                root = root->left;
            } else {
                root = s.top(); s.pop();
                root = root->right;
            }
        }
        return ans;
    }
};